plugins/modules/system/java_keystore.py

```diff
@@ -423,26 +469,50 @@ def create(self):
         else:
             self.openssl_create_pkcs12_bundle(keystore_p12_path)
 
+        if self.keystore_type == 'pkcs12':
+            # Preserve properties of the destination file, if any.
+            self.module.atomic_move(keystore_p12_path, self.keystore_path)
+            self.update_permissions()
+            self.result['changed'] = True
+            return self.result
+
         import_keystore_cmd = [self.keytool_bin, "-importkeystore",
                                "-destkeystore", self.keystore_path,
                                "-srckeystore", keystore_p12_path,
                                "-srcstoretype", "pkcs12",
                                "-alias", self.name,
                                "-noprompt"]
 
-        (rc, import_keystore_out, dummy) = self.module.run_command(
+        if self.keystore_type == 'jks':
+            keytool_help = self.module.run_command([self.keytool_bin, '-importkeystore', '-help'])
+            if '-deststoretype' in keytool_help[1] + keytool_help[2]:
+                import_keystore_cmd.insert(4, "-deststoretype")
+                import_keystore_cmd.insert(5, self.keystore_type)
+
+        keystore_backup = None
+        if self.exists():
+            keystore_backup = self.keystore_path + '.tmpbak'
+            self.module.add_cleanup_file(keystore_backup)
+            # Preserve properties of the source file
+            self.module.preserved_copy(self.keystore_path, keystore_backup)
+            os.remove(self.keystore_path)
```

Comment:
So if something causes the module to fail before the file is copied back, both the backup *and* the keystore are deleted?

Reply:
I don't see where. The file is copied back ~3 lines later if the next command (`module.run_command(keytool_command, check_rc=False)`) fails.

Unless there is something in `module.run_command` (with check_rc=False) or in `self.result = dict(...)` that could raise an exception, and `module.add_cleanup_file` cleanup the file in that case too... You think this may happen ?

Reply:
Done. The backup is now removed right after its restoration, or if keytool succeeds

Reply:
> I don't see where. The file is copied back ~3 lines later if the next command (`module.run_command(keytool_command, check_rc=False)`) fails.

A lot can happen in 3 lines, like a programming error resulting in an exception (can also be in ansible-core), some unexpected error in `run_command` (`OSError`, `IOError` and `Exception` raised by its main code result in it calling `fail_json`), ... While this is not so likely to happen, the result would be pretty ugly (both backup and original file deleted).

> Unless there is something in `module.run_command` (with check_rc=False) or in `self.result = dict(...)` that could raise an exception, and `module.add_cleanup_file` cleanup the file in that case too... You think this may happen ?

Maybe not extremely likely, but it can happen. The most likely case is probably that someone removed the keytool binary directly before the `run_command` call.